### ISSUE TITLE
Call to `z3.simplify` corrupts expression

### ISSUE THREAD

**Issue by rindPHI** [Z3Prover/z3#6043]
Dear Z3 team,

I am experiencing some strange behavior of `z3.simplify()`. If I do not misinterpret, e.g., the semantics of the conversion of signed bit vectors to integers, this could be a severe bug in Z3.

What happens is that *a call to z3.simplify corrupts the simplified expression.* The "simplified" expression suddenly contains sub-expressions that should not be there.

In my demonstrator below (for Z3 v4.8.17.0, but this also occurred for v4.8.14.0), I am converting a sequence of two (signed) bit vectors to a sequence of two integers. When I apply `simplify` to the results of the conversions of the elements, the resulting expressions also access the respectively different sequence element.

The bug only occurs for *signed* conversion and only if I access at least two list elements.

I hope that my illustrating example helps in diagnosing the issue. This behavior has cost me about three days in total, so I hope that I can at least contribute to fixing the bug or learning about what I did wrong.

Thanks!

Best,
Dominic

```python
from typing import Set
import z3


def compute(do_simplify=False) -> z3.SeqRef:
    result = z3.Empty(z3.SeqSort(z3.IntSort()))
    seq = z3.Const('seq', z3.SeqSort(z3.BitVecSort(8)))

    for idx in range(2):
        result_elem = z3.BV2Int(seq[z3.IntVal(idx)], is_signed=True)

        if do_simplify:
            result_elem = z3.simplify(result_elem)

        result = z3.Concat(result, z3.Unit(result_elem))

    return result


def accessed_indices(expr: z3.ExprRef) -> Set[int]:
    if expr.decl().name() == 'seq.nth_i':
        return {expr.children()[1].as_long()}

    return {
        index
        for child in expr.children()
        for index in accessed_indices(child)}


if __name__ == '__main__':
    assert z3.get_version() == (4, 8, 17, 0)

    result_1 = z3.simplify(compute(do_simplify=False)[1])
    result_2 = z3.simplify(compute(do_simplify=True)[1])

    print(result_1)
    print(result_2)

    assert accessed_indices(result_1) == {1}
    assert accessed_indices(result_2) == {1}  # Fails

```

**Comment by NikolajBjorner** [comment]
I don't think it is a bug. The simplifier just creates a larger expression that includes a conditional that uses index 0.
The conditional is not needed (can be simplified). 
I updated the simplifier to detect such situations so now it will pass your test.

**Comment by rindPHI** [comment]
Thanks for updating the simplifier!

Just one final remark: In my case, it manifested as a bug (in my concolic execution engine) since the Z3-produced models contained random-looking assignments for all accessed indices, while without these other indices in the expression, the sequence would be more stable. Also, it made a difference *where* in my code I applied `simplify` on the expression, and how many times (it sometimes was not idempotent). 

I actually checked the expressions for equivalence (add
```python
s = z3.Solver()
s.add(z3.Not(result_1 == result_2))
assert s.check() == z3.unsat
```
to the code above) and they are, as you say, equivalent, so in some sense you're right about this not being a bug. The wider implications, however, make it look like a bug in situations where you'd like "stable" models. Would be great if this worked now, thanks again!
